src/test/librust/sapling_wallet_tests.cpp

```diff
@@ -0,0 +1,1133 @@
+    setupWallet(wallet);
+
+    // Generate Sapling address
+    auto sk = GetTestMasterSaplingSpendingKey();
+    auto expsk = sk.expsk;
+    auto extfvk = sk.ToXFVK();
+    auto ivk = extfvk.fvk.in_viewing_key();
+    auto pk = sk.DefaultAddress();
+
+    BOOST_CHECK(wallet.AddSaplingZKey(sk));
+    BOOST_CHECK(wallet.HaveSaplingSpendingKey(extfvk));
+
+    // Generate note A
+    libzcash::SaplingNote note(pk, 50000);
+    auto cm = note.cmu().get();
+    SaplingMerkleTree saplingTree;
+    saplingTree.append(cm);
+    auto anchor = saplingTree.root();
+    auto witness = saplingTree.witness();
+
+    // Generate tx to create output note B
+    auto builder = TransactionBuilder(consensusParams, 1);
+    builder.AddSaplingSpend(expsk, note, anchor, witness);
+    builder.AddSaplingOutput(extfvk.fvk.ovk, pk, 35000, {});
+    auto tx = builder.Build().GetTxOrThrow();
+    CWalletTx wtx {&wallet, tx};
+
+    // Fake-mine the transaction
+    BOOST_CHECK_EQUAL(0, chainActive.Height());
+    CBlock block;
+    block.hashPrevBlock = chainActive[0]->GetBlockHash();
+    block.vtx.emplace_back(MakeTransactionRef(wtx));
+    block.hashMerkleRoot = BlockMerkleRoot(block);
+    auto blockHash = block.GetHash();
+    CBlockIndex fakeIndex {block};
+    BlockMap::iterator mi = mapBlockIndex.emplace(blockHash, &fakeIndex).first;
+    fakeIndex.phashBlock = &((*mi).first);
+    chainActive.SetTip(&fakeIndex);
+    BOOST_CHECK(chainActive.Contains(&fakeIndex));
+    BOOST_CHECK_EQUAL(0, chainActive.Height());
+
+    // Simulate SyncTransaction which calls AddToWalletIfInvolvingMe
+    auto saplingNoteData = wallet.GetSaplingScriptPubKeyMan()->FindMySaplingNotes(wtx).first;
+    BOOST_CHECK(saplingNoteData.size() > 0);
+    wtx.SetSaplingNoteData(saplingNoteData);
+    wtx.SetMerkleBranch(mapBlockIndex[blockHash], 0);
+    BOOST_CHECK(wallet.LoadToWallet(wtx));
+
+    // Simulate receiving new block and ChainTip signal
+    wallet.IncrementNoteWitnesses(&fakeIndex, &block, saplingTree);
+    wallet.GetSaplingScriptPubKeyMan()->UpdateSaplingNullifierNoteMapForBlock(&block);
+
+    // Retrieve the updated wtx from wallet
+    uint256 hash = wtx.GetHash();
+    wtx = wallet.mapWallet[hash];
+
+    // Decrypt output note B
+    auto maybe_pt = libzcash::SaplingNotePlaintext::decrypt(
+            wtx.sapData->vShieldedOutput[0].encCiphertext,
+            ivk,
+            wtx.sapData->vShieldedOutput[0].ephemeralKey,
+            wtx.sapData->vShieldedOutput[0].cmu);
+    BOOST_CHECK(static_cast<bool>(maybe_pt) == true);
+    auto maybe_note = maybe_pt.get().note(ivk);
+    BOOST_CHECK(static_cast<bool>(maybe_note) == true);
+    auto note2 = maybe_note.get();
+
+    SaplingOutPoint sop0(wtx.GetHash(), 0);
+    auto spend_note_witness =  wtx.mapSaplingNoteData[sop0].witnesses.front();
+    auto maybe_nf = note2.nullifier(extfvk.fvk, spend_note_witness.position());
+    BOOST_CHECK(static_cast<bool>(maybe_nf) == true);
+    auto nullifier2 = maybe_nf.get();
```

Comment:
(9cabc2ed2399221b23491c536254488dda44e60e)
It seems that  this variable is never used.